wandb/sdk/lib/filesystem.py

```diff
@@ -124,3 +127,64 @@ def copy_or_overwrite_changed(source_path: StrPath, target_path: StrPath) -> Str
         os.chmod(target_path, permissions_plus_write)
 
     return return_type(target_path)  # type: ignore  # 'os.PathLike' is abstract.
+
+
+@contextlib.contextmanager
+def safe_open(
+    path: StrPath, mode: str = "r", *args: Any, **kwargs: Any
+) -> Generator[IO, None, None]:
+    """Open a file, ensuring any changes only apply atomically after close.
+
+    This context manager ensures that even unsuccessful writes will not leave a "dirty"
+    file or overwrite good data, and that all temp data is cleaned up. Otherwise, the
+    semantics and behavior should be identical to the built-in open() function.
+    """
+    path = Path(path)
```

Comment:
Nit: people can cwd inside the context, let's remember the absolute path.
```suggestion
    path = Path(path).absolute()
```

Reply:
Good catch. I think it should actually be `resolve()` though, since we should also follow symlinks in order to ensure we're creating the temp file on the correct file system.
```suggestion
    path = Path(path).resolve()
```